src/priority-queue.ts

```diff
@@ -0,0 +1,262 @@
+import {ArmorCollectionSelector} from './selector';
+
+/**
+ * decide what to do when no comparator is passed
+ */
+
+export interface ArmorPriorityQueueOptions<T> {
+	elements?: Array<T>;
+	comparator?: (nodeOneIndex: number, nodeTwoIndex: number) => boolean;
+}
+
+const defaultComparator = function (nodeOneIndex: number, nodeTwoIndex: number) {
+	if (typeof nodeOneIndex !== 'number') {
+		return false;
+	}
+	if (typeof nodeTwoIndex !== 'number') {
+		return false;
+	}
+
+	return JSON.stringify(this.elements[nodeOneIndex]) < JSON.stringify(this.elements[nodeTwoIndex]);
+};
+
+export class ArmorPriorityQueue<T> implements ArmorCollection<T> {
+	public elements: T[];
+	public comparator: (nodeOneIndex: number, nodeTwoIndex: number) => boolean;
+
+	constructor(options: ArmorPriorityQueueOptions<T> = {}) {
+		if (typeof options.comparator === 'function') {
+			this.comparator = options.comparator;
+		} else {
+			this.comparator = defaultComparator;
+		}
+
+		if (options.elements === undefined) {
+			options.elements = [];
+		}
+		if (!Array.isArray(options.elements)) {
+			throw 'options.elements must be an array';
+		}
+
+		this.elements = [];
+		options.elements.forEach((element) => {
+			this.push(element);
+		});
+	}
+
+	public select(): ArmorCollectionSelector<T> {
+		const selector = new ArmorCollectionSelector<T>(this);
+
+		return selector;
+	}
+
+	public clear(): ArmorPriorityQueue<T> {
+		this.elements = [];
+
+		return this;
+	}
+
+	public size(): number {
+		if (!Array.isArray(this.elements)) {
+			return 0;
+		}
+		return this.elements.length;
+	}
+
+	public front(): T | null {
+		if (this.size() === 0) {
+			return null;
+		}
+		return this.elements[0];
+	}
+
+	public swapNodes(nodeOneIndex: number | null, nodeTwoIndex: number | null): void {
+		if (typeof nodeOneIndex !== 'number') {
+			return;
+		}
+		if (typeof nodeTwoIndex !== 'number') {
+			return;
+		}
+		if (nodeOneIndex < 0) {
+			return;
+		}
+		if (nodeTwoIndex < 0) {
+			return;
+		}
+		if (nodeOneIndex >= this.size()) {
+			return;
+		}
+		if (nodeTwoIndex >= this.size()) {
+			return;
+		}
+		if (nodeOneIndex === nodeTwoIndex) {
+			return;
+		}
+		if (nodeOneIndex % 1 !== 0) {
+			return;
+		}
+		if (nodeTwoIndex % 1 !== 0) {
+			return;
+		}
+
+		const nodeOneInfo = this.elements[nodeOneIndex];
+		this.elements[nodeOneIndex] = this.elements[nodeTwoIndex];
+		this.elements[nodeTwoIndex] = nodeOneInfo;
+	}
+
+	public getParentNodeIndex(nodeIndex: number | null): number | null {
+		if (typeof nodeIndex !== 'number') {
+			return null;
+		}
+		if (nodeIndex <= 0) {
+			return null;
+		}
+		if (nodeIndex >= this.size()) {
+			return null;
+		}
+		if (nodeIndex % 1 !== 0) {
+			return null;
+		}
+
+		const parentIndex = Math.floor((nodeIndex - 1) / 2);
+		if (parentIndex >= this.size()) {
+			return null;
+		}
+
+		return parentIndex;
+	}
+
+	public getChildNodesIndexes(nodeIndex: number | null): [number | null, number | null] {
```

Comment:
- [ ] Replace the array return with interface to clarifies how a caller uses the result.

There's no technical issues with returning an array. However, as a library used by third parties we want the signature to popup in the IDE and make usage as clear as possible. Returning a simple object isn't functionally much different than a 2 element array, but type name and property names make a huge difference in cognitive complexity.  I have to look at the function and figure out why it returns an array and how I might use that.

Here's an example type you can use:
```typescript
// this name is stupidly long
export interface ArmorPriorityQueueNodeChildren {
   left: number | null;
   right: number | null;
}
```
Then, after changing the function signature to return ArmorPriorityQueueNodeChildren<T> each array return becomes an object. You don't need to type the returns as typescript will check the type against the expected return type for you.
```typescript
// Return nulls
return { left: null, right: null };

// Return indices:
return { left: childOneIndex, right: childTwoIndex };
```